### PR TITLE
Allow zero WCSAxes ticks

### PR DESCRIPTION
**Pull request by dstansby**
Fixes https://github.com/astropy/astropy/issues/14159

### Description

This pull request is to address ...

Fixes #<Issue Number>

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label. Codestyle issues can be fixed by the [bot](https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#pre-commit).
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by dstansby** [comment]
🤷 - if you want to tag me in an issue for those new failures I might be able to take a look at them at some point.

File: astropy/visualization/wcsaxes/tests/test_images.py [hunk 1]
```diff
@@ -486,6 +486,18 @@ def test_ticks_labels(self):
...
+    def test_no_ticks(self):
+        # Check that setting no ticks works
+        fig = plt.figure(figsize=(6, 6))
+        ax = fig.add_axes(
+            [0.1, 0.1, 0.8, 0.8], projection=WCS(self.msx_header), aspect="equal"
+        )
+        ax.set_xlim(-0.5, 148.5)
+        ax.set_ylim(-0.5, 148.5)
+        ax.coords[0].set_ticks(number=0)
+        return fig
```
Comment (dhomeier) [inline astropy/visualization/wcsaxes/tests/test_images.py:499]:
```suggestion
        # Draw to trigger calling `_update_ticks`.
        plt.draw()
        return fig
```
Without enforcing a `draw_wcsaxes` call I don't get the test to raise without your fixes installed.

**Comment by dhomeier** [comment]
> 🤷 - if you want to tag me in an issue for those new failures I might be able to take a look at them at some point.

#14174

File: astropy/visualization/wcsaxes/tests/test_images.py (hunk 1 above)
Comment (dstansby) [reply astropy/visualization/wcsaxes/tests/test_images.py:499]: Hm, shouldn't `draw_wcsaxes()` be called somwhwere when the figure is rendered for the figure test? I think it's best to leave the `plt.draw()` call out for simplicity here. If any ticks are drawn, they'll show up in the figure test image.

Comment (dhomeier) [reply astropy/visualization/wcsaxes/tests/test_images.py:499]: That's right, when creating the comparison figures, the actual methods are called. But that's only tested if `mpl_image_compare` is installed.
I've copied the `draw` from other tests below, which will ensure to always fail if rendering is broken, but I don't know if there are rules when to test a method in the regular image tests.

**Comment by pllim** [comment]
CI should be green on main again. Please rebase. Thanks!

**Comment by astrofrog** [comment]
Looks like the hash is missing for Matplotlib 3.2.2?

**Comment by astrofrog** [comment]
I've updated the hash for 3.2.2 and rebased

**Review by astrofrog — APPROVED** [review]
Looks good to me know if the CI is happy!

**Comment by pllim** [comment]
I am not sure if this will backport cleanly because we bumped matplotlib minversion on main. Can we not backport?

**Comment by astrofrog** [comment]
I would like to backport

File: astropy/visualization/wcsaxes/coordinate_helpers.py [hunk 2]
```diff
@@ -957,6 +957,8 @@ def _update_grid_lines(self):
         tick_world_coordinates_values = tick_world_coordinates.to_value(self.coord_unit)
 
         n_coord = len(tick_world_coordinates_values)
+        if n_coord == 0:
```
[the remark this answers is not in the document]
Comment (astrofrog) [reply astropy/visualization/wcsaxes/coordinate_helpers.py:960]: Maybe the coverage results aren't getting uploaded from CircleCI?

[the remark this answers is not in the document]
Comment (astrofrog) [reply astropy/visualization/wcsaxes/coordinate_helpers.py:960]: This is for grid lines, which the test doesn't include, will see if I can fix.

[the remark this answers is not in the document]
Comment (astrofrog) [reply astropy/visualization/wcsaxes/coordinate_helpers.py:960]: Yes that was it, pushing a new commit

File: docs/changes/visualization/14160.bugfix.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+``CoordinateHelper.set_ticks()`` now accepts ``nticks=0``. Previously it errored.
```
Comment (pllim) [inline docs/changes/visualization/14160.bugfix.rst:1]: I think this is a typo?
```suggestion
``CoordinateHelper.set_ticks()`` now accepts ``number=0``. Previously it errored.
```

Comment (larrybradley) [reply docs/changes/visualization/14160.bugfix.rst:1]: Yeah, that appears to be a typo.

**Review by larrybradley — CHANGES_REQUESTED** [review]

**Comment by pllim** [comment]
This now also needs a rebase. We dropped mpl311 and mpl322. It is mpl334 now.

**Review by larrybradley — APPROVED** [review]

**Comment by larrybradley** [comment]
@pllim I rebased and fixed the changelog.

**Comment by pllim** [comment]
There is a failure https://output.circle-artifacts.com/output/job/afb29499-538d-46b1-b9b3-974414264680/artifacts/0/results/fig_comparison.html

File: astropy/tests/figures/py39-test-image-mpldev-cov.json [hunk 4]
```diff
@@ -21,6 +21,7 @@
   "astropy.visualization.wcsaxes.tests.test_images.TestBasic.test_changed_axis_units": "056ec7123c67fa7ee6269bec47641bfe4932630a79d02029e71164b522fadacd",
   "astropy.visualization.wcsaxes.tests.test_images.TestBasic.test_minor_ticks": "5224f8fa725901ff74a4558987ceaae0363f98894d2c6a12d8c557db500ff7fe",
   "astropy.visualization.wcsaxes.tests.test_images.TestBasic.test_ticks_labels": "f4944296c37a6e8ea9d30126709da3567a25908c1665e8cad10e894dd87cd54d",
+  "astropy.visualization.wcsaxes.tests.test_images.TestBasic.test_no_ticks": "14c20201942f1db9796db8747ffcdec77254d79a9ae7dc45ce668ebf07c6de8e",
```
Comment (pllim) [inline astropy/tests/figures/py39-test-image-mpldev-cov.json:24]: CircleCI failed with `Baseline hash not found` but can we ignore it?

**Comment by astrofrog** [comment]
@pllim - no the issue is that we are missing an update to the mpl334 file. If no one gets to it by later today I can try and fix that.

**Comment by pllim** [comment]
I don't know exactly what that means, so I'll leave it to you then; thanks, @astrofrog !

**Comment by dhomeier** [comment]
Some mismatches with the hashes:
```
__________________________ TestBasic.test_tick_angles __________________________
Hash 475f028bd48a51fe9dfc07fe3d58e63e629511fffeb3a4db27a26f6427b4256b doesn't match hash e828e41d48899dd706cc58bb6fb93427d30b4c16ccd360813b391a880bee692c in library /home/circleci/project/astropy/tests/figures/py39-test-image-mpl334-cov.json for test astropy.visualization.wcsaxes.tests.test_images.TestBasic.test_tick_angles.

Image comparison test
---------------------
The comparison to the baseline image succeeded.
__________________ TestBasic.test_tick_angles_non_square_axes __________________
Hash 3a3a15f71e7303368bd7cebb70c6fd1f9d6710edf8af2ceb68642c64c74492a2 doesn't match hash adc79cf700df608e662bd9b4904fd8160dce15e67d04698c04f650808411ea97 in library /home/circleci/project/astropy/tests/figures/py39-test-image-mpl334-cov.json for test astropy.visualization.wcsaxes.tests.test_images.TestBasic.test_tick_angles_non_square_axes.

Image comparison test
---------------------
The comparison to the baseline image succeeded.
___________________________ test_tickable_gridlines ____________________________
Hash 1be26963aaba221f9cd396aea0b9618e05ab8a49f8c02c0b4540041e91d06ce7 doesn't match hash ba289534a6a65f01d4dbe050bdcab5e97f19549339896dde84e3b861cd1b07ba in library /home/circleci/project/astropy/tests/figures/py39-test-image-mpl334-cov.json for test astropy.visualization.wcsaxes.tests.test_images.test_tickable_gridlines.

Image comparison test
---------------------
The comparison to the baseline image succeeded.
__________________ TestTransformCoordMeta.test_coords_overlay __________________
Hash f84074e12142676d17aca8e907065cf1eec4a55ea7d60ce7d7f22f456c09c24c doesn't match hash 673e604c362ea09ff027da86b5c7f39bf41469dc92143b382010144ab5779330 in library /home/circleci/project/astropy/tests/figures/py39-test-image-mpl334-cov.json for test astropy.visualization.wcsaxes.tests.test_transform_coord_meta.TestTransformCoordMeta.test_coords_overlay.

Image comparison test
---------------------
The comparison to the baseline image succeeded.
___________________ TestTransformCoordMeta.test_direct_init ____________________
Hash 99e84ea7ae80a9444c73d4fb9f24d707795149e1aeb18c3c1a7072bc3e134732 doesn't match hash 3d7c5f4e7a2e9025b9f05001cd41b24a3f3a09420adb5ea2241dd82188cccff2 in library /home/circleci/project/astropy/tests/figures/py39-test-image-mpl334-cov.json for test astropy.visualization.wcsaxes.tests.test_transform_coord_meta.TestTransformCoordMeta.test_direct_init.
```

**Comment by pllim** [comment]
Do we have to backport?

**Pull request merged**